feat(bridge): OpenID flow (#1006)
keptn/keptn.github.io: content/docs/0.11.x/reference/bridge/oauth/index.md

- `clientID` - Client ID, provided by the identity provider.
- `clientSecret` (optional) - Client secret, provided by the identity provider.
- `IDTokenAlg` (optional) - Algorithm that is used to verify the ID token (e.g. `ES256`). Default is `RS256`.
- `scope` (optional) - Additional scopes that should be added to the authentication flow (e.g. `profile email`), separated by space.
- `userIdentifier` (optional) - The property of the ID token that identifies the user. Default is `name` and fallback to `nickname`, `preferred_username` and `email`.

## Authentication flow with the identity provider

The following diagram shows the expected authentication flow with a custom identity provider that follows the implementation details explained below. 

{{< popup_image
link="./assets/oauth-flow.png"
caption="Authentication flow with identity provider" width="800px">}}

Keptn Bridge should check the existence of the identity provider and also validate responses obtained from it. For example, next are requests and responses the Keptn Bridge can expect from the identity provider.

*To obtain discovery request:*

```
Request :

>    Get <identity_provider>/.well-known/openid-configuration

Reviewer: ```suggestion
   GET <identity_provider>/.well-known/openid-configuration
```

```
Request :

   POST <token_endpoint>

   {
     "state" : <state>,
     "code" : <authorization_code>,
     "scope": "openid",
     "client_id": <client_id>,
     "client_secret": <client_secret>, 
     "nonce": <nonce>,
     "code_verifier": <code_verifier>
   }

Response :

  {
   "id_token": <id_token>,
   "access_token": <access_token>,
   "refresh_token": <refresh_token>,
   "expires_in": <expires_in>
  }
```

> The identity provider must implement logic to generate authorization requests, obtain tokens for valid authorization codes and validate permission checks based on token contents or user information. Furthermore, authorization request and token request should be linked ideally through state, nonce, code challenge and code verifier. The identity provider is expected to follow recommended best practices such as use of [PKCE](https://tools.ietf.org/html/rfc7636) and JWK validations where applicable.

Reviewer 2: Is `nonce` a typo?

Author: No https://openid.net/specs/openid-connect-core-1_0.html#:~:text=auth_time%20response%20parameter.)-,nonce,-String%20value%20used

```
Request :

   POST <token_endpoint>

   {
     "state" : <state>,
     "code" : <authorization_code>,
     "scope": "openid",
     "client_id": <client_id>,
     "client_secret": <client_secret>, 
     "nonce": <nonce>,
     "code_verifier": <code_verifier>
   }

Response :

  {
   "id_token": <id_token>,
   "access_token": <access_token>,
   "refresh_token": <refresh_token>,
   "expires_in": <expires_in>
  }
```

> The identity provider must implement logic to generate authorization requests, obtain tokens for valid authorization codes and validate permission checks based on token contents or user information. Furthermore, authorization request and token request should be linked ideally through state, nonce, code challenge and code verifier. The identity provider is expected to follow recommended best practices such as use of [PKCE](https://tools.ietf.org/html/rfc7636) and JWK validations where applicable.

Reviewer: ```suggestion
The identity provider must implement logic to:
- generate authorization requests,
- obtain tokens for valid authorization codes, and
- validate permission checks based on token contents or user information. 
Furthermore, authorization requests and token requests should be linked ideally through state, nonce, code challenge, and code verifier. The identity provider is expected to follow recommended best practices such as [PKCE](https://tools.ietf.org/html/rfc7636) and JWK validations where applicable.
```

> The Keptn Bridge expects the environment variable `OAUTH_DISCOVERY`. This must direct to a discovery endpoint with the following details:

Reviewer: ```suggestion
When enabled, the Keptn Bridge expects the environment variable `OAUTH_DISCOVERY` to be correctly set. This must direct to a discovery endpoint with the following details:
```

1. Endpoint for authorization request generation - `authorization_endpopint`.
2. Endpoint to handle tokens and provide the login decision - `token_endpoint`.
3. Endpoint to fetch JSON Web key set to validate the ID token - `jwks_uri`.
4. (optional) Endpoint to handle the logout - `end_session_endpoint`.

With above-mentioned keys, the following is a sample response of the discovery

```

> Get http://identity-provider:8080/.well-known/openid-configuration

Reviewer: ```suggestion
GET http://identity-provider:8080/.well-known/openid-configuration
```

{
 "authorization_endpoint": "http://identity-provider:8080/authorization",
 "token_endpoint": "http://identity-provider:8080/token_decision"
 "jwks_uri": "http://identity-provider:8080/jwks"
}
```

> From the response, Keptn bridge identifies the specific endpoints it needs to consume. Following sections provide specific details of these endpoints.

Reviewer: ```suggestion
From the response, Keptn bridge identifies the specific endpoints it needs to consume. The following sections provide specific details of these endpoints.
```

### authorization endpoint

> Generates an authentication code and redirects to the given redirect URI.

Reviewer: ```suggestion
It generates an authentication code and redirects to the given redirect URI.
```

The following query parameters must be supported:
- client_id
- redirect_uri
- scope
- state
- nonce
- code_challenge
- code_challenge_method

### token endpoint

> Consumes code and state of the authorization response and provides the login decision. For simplicity, following are the currently supported decisions:

Reviewer: ```suggestion
It consumes the code challenge and state of the authorization response and provides the login decision. For simplicity, the following are the currently supported decisions:
```

Author: It's the `authorization code`, not the `code challenge`. The `code challenge` is sent to the authorization endpoint, not to the token endpoint.

Reviewer: Thank you for the clarification! :)

Successful login example:

```

> Post http://identity-provider:8080/token_decision

Reviewer: ```suggestion
POST http://identity-provider:8080/token_decision
```

HTTP 200 OK
Content-Type: application/json
{
  "id_token": <id_token>,
  "access_token": <access_token>,
  "refresh_token": <refresh_token>,
  "expires_in": <expires_in>
}
```

Unsuccessful login example:

```

> Post http://identity-provider:8080/token_decision

Reviewer: ```suggestion
POST http://identity-provider:8080/token_decision
```

{
  "message": "User Alex does not have permission to login to Keptn."
}
```

## OpenID Connect via Microsoft

> To set up SSO via Microsoft you have to [register an application](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) in order to get a client id, client secret and discovery endpoint.\

Reviewer: ```suggestion
To set up SSO via Microsoft you have to [register an application](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) in order to get a client id, client secret, and a discovery endpoint.
```

> Then the following environment variables can be set when installing keptn

Reviewer: ```suggestion
Then the following environment variables can be set when installing Keptn
```

```
bridge:
  ...
  oauth:
    enabled: true
    discovery: "https://login.microsoftonline.com/${directory_tenant_id}/v2.0/.well-known/openid-configuration"
    secureCookie: true
    baseUrl: <base_url>
    clientID: <client_id>
    clientSecret: <client_secret>
    scope: "email"
```

> or later directly for the bridge server. After entering new environment variables, the server has to be restarted.

Reviewer: ```suggestion
It is also possible to directly change the `Deployment` manifest of an existing Keptn installation. After entering the new environment values, the Bridge pod has to be restarted.
```

```
OAUTH_ENABLED: "true"
OAUTH_DISCOVERY: "https://login.microsoftonline.com/${directory_tenant_id}/v2.0/.well-known/openid-configuration"
SECURE_COOKIE: "true"
OAUTH_BASE_URL: <base_url>
OAUTH_CLIENT_ID: <client_id>
OAUTH_CLIENT_SECRET: <client_secret>
OAUTH_SCOPE: "email"
```

> When accessing the bridge, the user is redirected to the identity provider.

Reviewer: ```suggestion
When accessing the Bridge, the user is redirected to the identity provider.
```

{{< popup_image
link="./assets/oauth-login-message.png"
caption="Accessing bridge without being logged in" width="800px">}}
{{< popup_image
link="./assets/oauth-login.png"
caption="Entering user credentials" width="400px">}}

After the user successfully logs in with his Microsoft credentials, he is redirected back to the bridge.

> Once redirected, the bridge server fetches the user tokens and creates a session. The user is now successfully logged in.

Reviewer: ```suggestion
Once redirected, the Bridge fetches the user tokens and creates a session. The user is now successfully logged in.
```
